File: app/crawler/robots.py
"""robots.txt handling per RFC 9309.

`urllib.robotparser` is not used because it does plain prefix matching and ignores the
`*` / `$` wildcards, which would silently crawl paths a site has disallowed.

Rules:
- The group whose user-agent matches our product token is used (merged if repeated);
  otherwise the `*` group; otherwise everything is allowed.
- The most specific (longest) matching rule wins; `Allow` wins ties.
- robots.txt fetch: 2xx -> parse, 4xx -> allow all, 5xx / network failure -> disallow all.
"""

from __future__ import annotations

import re
from contextlib import suppress
from dataclasses import dataclass, field
from urllib.parse import unquote, urlsplit


@dataclass
class _Group:
    agents: list[str] = field(default_factory=list)
    rules: list[tuple[bool, str]] = field(default_factory=list)  # (allow, pattern)
    crawl_delay: float | None = None


@dataclass
class RobotsRules:
    rules: list[tuple[bool, str]] = field(default_factory=list)
    crawl_delay: float | None = None
    sitemaps: list[str] = field(default_factory=list)
    allow_all: bool = False
    disallow_all: bool = False
    source_status: int | None = None

    @classmethod
    def allowing_all(cls, status: int | None = None) -> RobotsRules:
        return cls(allow_all=True, source_status=status)

    @classmethod
    def disallowing_all(cls, status: int | None = None) -> RobotsRules:
        return cls(disallow_all=True, source_status=status)

    def can_fetch(self, url: str) -> bool:
        parts = urlsplit(url)
        path = parts.path or "/"
        if path == "/robots.txt":
            return True
        if self.disallow_all:
            return False
        if self.allow_all or not self.rules:
            return True
        target = _normalise_path(path + (f"?{parts.query}" if parts.query else ""))
        best: tuple[int, bool] | None = None  # (specificity, allow)
        for allow, pattern in self.rules:
            if _matches(pattern, target):
                candidate = (len(pattern), allow)
                if best is None or candidate[0] > best[0] or (candidate[0] == best[0] and allow):
                    best = candidate
        return True if best is None else best[1]


def _normalise_path(value: str) -> str:
    # Compare in decoded form so /a%2Db and /a-b are treated alike.
    return unquote(value)


def _matches(pattern: str, path: str) -> bool:
    anchored = pattern.endswith("$")
    body = pattern[:-1] if anchored else pattern
    regex = ".*".join(re.escape(_normalise_path(piece)) for piece in body.split("*"))
    return re.match(regex + ("$" if anchored else ""), path) is not None


def product_token(user_agent: str) -> str:
    return user_agent.split("/", 1)[0].strip().lower()


def parse_robots(text: str, user_agent: str) -> RobotsRules:
    token = product_token(user_agent)
    groups: list[_Group] = []
    sitemaps: list[str] = []
    current: _Group | None = None
    last_was_agent = False
    for raw_line in text.splitlines():
        line = raw_line.split("#", 1)[0].strip()
        if not line or ":" not in line:
            continue
        key, _, value = line.partition(":")
        key, value = key.strip().lower(), value.strip()
        if key == "sitemap":
            if value:
                sitemaps.append(value)
            continue
        if key == "user-agent":
            if current is None or not last_was_agent:
                current = _Group()
                groups.append(current)
            current.agents.append(value.lower())
            last_was_agent = True
            continue
        last_was_agent = False
        if current is None:
            continue  # rules before any user-agent line are ignored
        if key in ("allow", "disallow"):
            if value:  # empty Disallow means "allow everything"
                current.rules.append((key == "allow", value))
        elif key == "crawl-delay":
            with suppress(ValueError):
                current.crawl_delay = float(value)

    specific = [g for g in groups if token in g.agents]
    chosen = specific or [g for g in groups if "*" in g.agents]
    rules = [rule for g in chosen for rule in g.rules]
    delays = [g.crawl_delay for g in chosen if g.crawl_delay is not None]
    return RobotsRules(
        rules=rules,
        crawl_delay=max(delays) if delays else None,
        sitemaps=list(dict.fromkeys(sitemaps)),
        allow_all=not chosen,
        source_status=200,
    )

File: src/crawler/robots.js
/**
 * robots.txt handling per RFC 9309 (app/crawler/robots.py).
 *
 * Rules:
 * - The group whose user-agent matches our product token is used (merged if repeated);
 *   otherwise the `*` group; otherwise everything is allowed.
 * - The most specific (longest) matching rule wins; `Allow` wins ties.
 * - robots.txt fetch: 2xx -> parse, 4xx -> allow all, 5xx / network failure -> disallow all.
 * - `*` and `$` wildcards are supported (plain prefix matching would silently crawl paths a
 *   site has disallowed).
 */

import { unquote, urlsplit } from '../utils/pyurl.js';
import { pyLen, pyLower, pyStrip, splitlines } from '../utils/pytext.js';

export class RobotsRules {
  constructor({ rules = [], crawlDelay = null, sitemaps = [], allowAll = false, disallowAll = false, sourceStatus = null } = {}) {
    this.rules = rules; // [allow: boolean, pattern: string]
    this.crawlDelay = crawlDelay;
    this.sitemaps = sitemaps;
    this.allowAll = allowAll;
    this.disallowAll = disallowAll;
    this.sourceStatus = sourceStatus;
  }

  static allowingAll(status = null) {
    return new RobotsRules({ allowAll: true, sourceStatus: status });
  }

  static disallowingAll(status = null) {
    return new RobotsRules({ disallowAll: true, sourceStatus: status });
  }

  canFetch(url) {
    const parts = urlsplit(url);
    const path = parts.path || '/';
    if (path === '/robots.txt') return true;
    if (this.disallowAll) return false;
    if (this.allowAll || !this.rules.length) return true;
    const target = normalisePath(path + (parts.query ? `?${parts.query}` : ''));
    let best = null; // [specificity, allow]
    for (const [allow, pattern] of this.rules) {
      if (matches(pattern, target)) {
        const specificity = pyLen(pattern);
        if (best === null || specificity > best[0] || (specificity === best[0] && allow)) best = [specificity, allow];
      }
    }
    return best === null ? true : best[1];
  }
}

// Compare in decoded form so /a%2Db and /a-b are treated alike.
function normalisePath(value) {
  return unquote(value);
}

const escapeRegExp = (value) => value.replace(/[.*+?^${}()|[\]\\/]/g, '\\$&');

function matches(pattern, path) {
  const anchored = pattern.endsWith('$');
  const body = anchored ? pattern.slice(0, -1) : pattern;
  const regex = body.split('*').map((piece) => escapeRegExp(normalisePath(piece))).join('.*');
  // Python `re.match`: anchored at the start; `.` does not match newlines.
  return new RegExp(`^(?:${regex})${anchored ? '$' : ''}`, 'u').test(path);
}

export function productToken(userAgent) {
  return pyLower(pyStrip(userAgent.split('/', 1)[0]));
}

/** Python `float()` for robots values ("2", "2.5", "1e3", "inf", "nan", "1_0"). */
function pyFloat(value) {
  const v = pyStrip(value);
  if (/^[+-]?(inf|infinity)$/i.test(v)) return v.startsWith('-') ? -Infinity : Infinity;
  if (/^[+-]?nan$/i.test(v)) return Number.NaN;
  if (!/^[+-]?(\d(_?\d)*(\.(\d(_?\d)*)?)?|\.\d(_?\d)*)([eE][+-]?\d(_?\d)*)?$/.test(v)) return null;
  return Number(v.replaceAll('_', ''));
}

/** Python `max()` over floats (NaN comparisons are always false, first wins on ties). */
function pyMax(values) {
  let best = values[0];
  for (const v of values.slice(1)) if (v > best) best = v;
  return best;
}

export function parseRobots(text, userAgent) {
  const token = productToken(userAgent);
  const groups = [];
  const sitemaps = [];
  let current = null;
  let lastWasAgent = false;
  for (const rawLine of splitlines(text)) {
    const line = pyStrip(rawLine.split('#', 1)[0]);
    if (!line || !line.includes(':')) continue;
    const idx = line.indexOf(':');
    const key = pyLower(pyStrip(line.slice(0, idx)));
    const value = pyStrip(line.slice(idx + 1));
    if (key === 'sitemap') {
      if (value) sitemaps.push(value);
      continue;
    }
    if (key === 'user-agent') {
      if (current === null || !lastWasAgent) {
        current = { agents: [], rules: [], crawlDelay: null };
        groups.push(current);
      }
      current.agents.push(pyLower(value));
      lastWasAgent = true;
      continue;
    }
    lastWasAgent = false;
    if (current === null) continue; // rules before any user-agent line are ignored
    if (key === 'allow' || key === 'disallow') {
      if (value) current.rules.push([key === 'allow', value]); // empty Disallow allows everything
    } else if (key === 'crawl-delay') {
      const delay = pyFloat(value);
      if (delay !== null) current.crawlDelay = delay;
    }
  }
  const specific = groups.filter((g) => g.agents.includes(token));
  const chosen = specific.length ? specific : groups.filter((g) => g.agents.includes('*'));
  const delays = chosen.map((g) => g.crawlDelay).filter((d) => d !== null);
  return new RobotsRules({
    rules: chosen.flatMap((g) => g.rules),
    crawlDelay: delays.length ? pyMax(delays) : null,
    sitemaps: [...new Set(sitemaps)],
    allowAll: !chosen.length,
    sourceStatus: 200,
  });
}
